File: programmers/도둑질.cpp
#include <string>
#include <vector>
#include<algorithm>
#include<cstring>

using namespace std;
int cache[1000000];
vector<int> Gmoney;

int cal(int n,bool first)
{
    if(first == false && Gmoney.size() <= n) return 0;
    if(first == true && Gmoney.size()-1 <= n) return 0;
    int& result = cache[n];
    if(result != -1) return result;
    result = max(cal(n+1,first),cal(n+2,first)+Gmoney[n]);
    return result;
}

int solution(vector<int> money) {
    int answer = 0;
    for(int i =0;i<money.size();i++)
    {
        cache[i] = -1;
    }
    Gmoney = money;
    answer = cal(2,true)+money[0];
    for(int i =0;i<money.size();i++)
    {
        cache[i] = -1;
    }
    answer = max(answer,cal(1,false));
    return answer;
}

File: JAVA/Programmers/도둑질.java
class Solution {
    static int dp1[];
    static int dp2[];
    public int solution(int[] money) {
        int answer = 0;
        dp1 = new int[money.length]; // 첫번째 털음 , 마지막 못텀
        dp2 = new int[money.length]; // 첫번쨰 안텀
        dp1[0] = money[0];
        dp1[1] = Math.max(money[0],money[1]);
        dp2[1] = money[1];
        int size = money.length;
        for(int i = 2 ; i < size;i++){
            if(i != size-1)
                dp1[i] = Math.max(dp1[i-1],dp1[i-2]+ money[i]) ;
            dp2[i] = Math.max(dp2[i-1],dp2[i-2]+ money[i]) ;
        }
        answer = Math.max(dp1[size-2],dp2[size-1]);
        if(size == 3){
            answer = Math.max(answer,money[2]);
        }
        return answer;
    }
}
